### ISSUE TITLE
For whatever reason eval being executed in production mode

### ISSUE THREAD

**Issue by solomax** [mathjax/MathJax-src#818]
TL;DR;

for whatever reason this https://github.com/mathjax/MathJax-src/blob/master/ts/components/version.ts#L37 code being executed at my app :(

Long version:
We are using MathJax to draw beautiful formulas at Apache OpenMeetings: https://github.com/apache/openmeetings/blob/master/openmeetings-web/src/main/front/wb/package.json#L22
And `browserify` as build tool

We are using some sort node-like configuration https://github.com/apache/openmeetings/blob/master/openmeetings-web/src/main/front/wb/src/wb-tool-stat-math.js#L4

Everything was working as expected (and still working in DEV mode)
BUT for whatever reason in PRODUCTION `eval` is being called to determine the MathJax version :(

I've hacked this by adding `window.PACKAGE_VERSION = '3.2.1';` to my source code
But this a bit dirty :(

Maybe it would be possible to introduce some eval-free mode?
Or maybe I'm doing something wrong?
Or maybe it would be possible to store version in some pre/post build step? :)))

Thanks in advance for the help :)

**Comment by dpvc** [comment]
MathJax uses the webpack DefinePlugin to replace `PACKAGE_VERSION` by the actual version:

https://github.com/mathjax/MathJax-src/blob/9f588aaf7ed8e7f38bfb636157bf29a94ee4804d/components/webpack.common.js#L49-L58

(so web components will have the value) and tries to look it up at run time otherwise (for node applications).  So if you are using some other package manage, you will need to do something similar.  It looks like maybe the [browserify-replace](https://www.npmjs.com/package/browserify-replace) package might be able to do the job for you.

Sorry that this update has had this negative impact for you.

**Comment by solomax** [comment]
Thanks!
was able to properly work-around this :)

**Comment by ryangr-texthelp** [comment]
Just to chip in here, we have also run into this issue.

Unfortunately, for us, we are using `mathjax-full` as a `dependency` for our library so we are not bundling it directly. Consumers of our library may use a bundler, but it would be difficult for us to try and enforce this as part of a consumer build process.

I understand the need for `PACKAGE_VERSION` to be defined, and it's great that it happens as part of the build process, but equally it can be hardcoded in source and overwritten at runtime if required.

If we go down this route, we're going to have to have separate instructions for Rollup, Webpack, Browserify, and so on.... simply for reading the version number.

**Comment by ShaMan123** [comment]
I haven't had time to deep into this but I agree with @ryangr-texthelp 
I maintain a library that depends on mathjax as well.
(@solomax funny seeing you here)

**Comment by solomax** [comment]
@ShaMan123 :)
We are using both both fabric.js and MathJax
It was fun to make fabric to draw beautiful formulas :)

**Comment by ShaMan123** [comment]
> It was fun to make fabric to draw beautiful formulas :)

Nice. I do it as well in a project.
BTW I see you use a whiteout functionality. Take a look at [EraserBrush](http://fabricjs.com/erasing).

**Comment by NicolasCARPi** [comment]
Hello,

Could this issue be reopened please?

I fully agree with @ryangr-texthelp's message: the current approach to get a simple `VERSION` string is flawed and IMHO over-complicated. Now don't get me wrong, @dpvc had good reasons to do it like this I'm sure. But in the end many users spent way too much time trying to figure out the correct way to feed the PACKAGE_VERSION to the lib. I ended up overwriting the source code to hardcode it before bundling it.

I believe the issue is with projects building mathjax from source. For some reason I couldn't make the DefinePlugin approach work (I'm using webpack). As Ryan said, this won't work down the line for many users. Or even in 2 years when webpack gets replaced ;p

It would be great if this issue could be adressed in the next patch. :)

Best,
~Nico

**Comment by NicolasCARPi** [comment]
Hey @dpvc.

I'm glad to read that. Hopefully I can remove that ugly script of mine to hardcode the version soon :)

I also agree that a patch release is necessary. If [this](https://github.com/mathjax/MathJax-src/pull/820) could be integrated, it would be great too!

Best,
~Nico

**Comment by solomax** [comment]
Thanks for addressing this @dpvc  :))

**Comment by dpvc** [comment]
I've made a pull request that should resolve this issue.  It creates the version.ts file using the package.json version number during the travis-ci publish script, so the version will just be a constant, as requested.

**Comment by NicolasCARPi** [comment]
Great, thanks @dpvc !

**Comment by solomax** [comment]
Just have updated to the latest release `3.2.2`
Works as expected without _hacks_ :)
Thanks for the fix!

**Comment by dpvc** [comment]
@solomax, thanks for the confirmation.  Glad it works better for you.

**Comment by NicolasCARPi** [comment]
Yes, can confirm here too :+1: Cheers!

**Comment by dpvc** [comment]
@NicolasCARPi, great thanks!
